sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamIOPushDownRule.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Calc;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.RelFactories;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.type.RelDataType;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.type.RelRecordType;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexCall;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexInputRef;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLiteral;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLocalRef;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexProgram;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.tools.RelBuilder;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.util.Pair;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+
+public class BeamIOPushDownRule extends RelOptRule {
+  // ~ Static fields/initializers ---------------------------------------------
+
+  public static final BeamIOPushDownRule INSTANCE =
+      new BeamIOPushDownRule(RelFactories.LOGICAL_BUILDER);
+
+  // ~ Constructors -----------------------------------------------------------
+
+  public BeamIOPushDownRule(RelBuilderFactory relBuilderFactory) {
+    super(operand(Calc.class, operand(BeamIOSourceRel.class, any())), relBuilderFactory, null);
+  }
+
+  // ~ Methods ----------------------------------------------------------------
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final BeamIOSourceRel ioSourceRel = call.rel(1);
+    final BeamSqlTable beamSqlTable = ioSourceRel.getBeamSqlTable();
+
+    if (!beamSqlTable.supportsProjects()) {
+      return;
+    }
+
+    // Nested rows are not supported at the moment
+    for (RelDataTypeField field : ioSourceRel.getRowType().getFieldList()) {
+      if (field.getType() instanceof RelRecordType) {
+        return;
+      }
+    }
+
+    final Calc calc = call.rel(0);
+    final RexProgram program = calc.getProgram();
+    final Pair<ImmutableList<RexNode>, ImmutableList<RexNode>> projectFilter = program.split();
+    final RelDataType calcInputRowType = program.getInputRowType();
+
+    // Find all input refs used by projects
+    Set<String> usedFields = new HashSet<>();
+    for (RexNode project : projectFilter.left) {
+      findUtilizedInputRefs(calcInputRowType, project, usedFields);
+    }
+
+    // Check if the calc can be dropped
+    if (isProjectRenameOnlyProgram(program)) {
+      call.transformTo(ioSourceRel.copy(calc.getRowType(), new ArrayList<>(usedFields)));
```

Comment:
Just noticed you have the same cast to ArrayList problem here. Is there a place you can get the fields properly sorted?

Reply:
Good catch! Moved this check to happen after generating new schema. Check for the most optimal case need to move as well.